jax/interpreters/xla.py

```diff
@@ -1075,13 +1075,14 @@ def __iter__(self):
     if self.ndim == 0:
       raise TypeError("iteration over a 0-d array")  # same as numpy error
     else:
-      return self._value.__iter__()
+      device = self.device_buffer.device()
+      if device is None or device.platform == 'cpu':
+        return iter(self._value)
```

Comment:
If this still returns NumPy arrays, I suspect that might be undesirable because of, e.g., different promotion semantics.

However, it might make sense in the short term until some of the overheads improve.

Reply:
Yeah, good point. We should revise this case so that it returns CPU DeviceArrays, maybe.

Reply:
On CPU, with `return iter(self._value)`, running the big timing script in the OP using the `big` array, we get:

```
time to grab first element, uncompiled: 0.002505779266357422 ms
time to grab first element, compiled: 0.0018811225891113281 ms
total time to grab all elements, uncompiled: 0.696418285369873 ms
average time to grab each element, uncompiled: 6.96418285369873e-05 ms
total time to grab all elements, compiled: 0.6600356101989746 ms
average time to grab each element, compiled: 6.600356101989746e-05 ms
```

If we instead keep doing the slicing in NumPy but device_put each result, i.e. we do something like this:

```python
      device = self.device_buffer.device()
      if device is None or device.platform == 'cpu':
        # do the slicing in NumPy for better performance
        return iter(self._value)
        device = device or jax.devices('cpu')
        aval = ShapedArray(self.aval.shape[1:], self.aval.dtype,
                           self.aval.weak_type)
        lexpr = lazy.array(aval.shape)
        return (make_device_array(aval, device, lexpr, *device_put(x, device))
                for x in self._value)
```

Then we get this timing:
```
time to grab first element, uncompiled: 0.01703500747680664 ms
time to grab first element, compiled: 0.015168190002441406 ms
total time to grab all elements, uncompiled: 79.03305768966675 ms
average time to grab each element, uncompiled: 0.007903305768966674 ms
total time to grab all elements, compiled: 79.05462265014648 ms
average time to grab each element, compiled: 0.007905462265014648 ms
```

So a 100x slowdown to return DeviceArrays rather than ndarrays on CPU. Perhaps we'll be able to speed that up soon, but I wanted to bounce this off you anyway. WDYT?

Reply:
Updated timings, being slightly less dumb:

```
time to grab first element, uncompiled: 0.01026153564453125 ms
time to grab first element, compiled: 0.009059906005859375 ms
total time to grab all elements, uncompiled: 42.66105651855469 ms
average time to grab each element, uncompiled: 0.004266105651855468 ms
total time to grab all elements, compiled: 42.526516914367676 ms
average time to grab each element, compiled: 0.004252651691436768 ms
```

Reply:
New times, after jit dispatch time upgrades in jaxlib==0.1.65 (remember these are noisy!):

```
time to grab first element, uncompiled: 0.013675689697265625 ms
time to grab first element, compiled: 0.012180805206298828 ms
total time to grab all elements, uncompiled: 38.03548336029053 ms
average time to grab each element, uncompiled: 0.003803548336029053 ms
total time to grab all elements, compiled: 37.87814140319824 ms
average time to grab each element, compiled: 0.0037878141403198244 ms
```